samtranslator/model/api/api_generator.py

```diff
@@ -208,6 +216,113 @@ def _add_cors(self):
+        swagger_editor = SwaggerEditor(self.definition_body)
+        auth_properties = AuthProperties(**self.auth)
+        authorizers = self._get_authorizers(auth_properties.Authorizers)
+
+        if authorizers:
+            swagger_editor.add_authorizers(authorizers)
+            self._set_default_authorizer(swagger_editor, authorizers, auth_properties.DefaultAuthorizer)
+
+        # Assign the Swagger back to template
+        self.definition_body = swagger_editor.swagger
+
+    def _get_authorizers(self, authorizers_config):
+        if not authorizers_config:
+            return None
+
+        if not isinstance(authorizers_config, dict):
+            raise InvalidResourceException(self.logical_id,
+                                           "Authorizers must be a dictionary")
+        authorizers = {}
+
+        for authorizer_name, authorizer in authorizers_config.items():
+            authorizers[authorizer_name] = ApiGatewayAuthorizer(
+                api_logical_id=self.logical_id,
+                name=authorizer_name,
+                user_pool_arn=authorizer.get('UserPoolArn'),
+                function_arn=authorizer.get('FunctionArn'),
+                identity=authorizer.get('Identity'),
+                function_payload_type=authorizer.get('FunctionPayloadType'),
+                function_invoke_role=authorizer.get('FunctionInvokeRole')
+            )
+
+        return authorizers
+
+    def _get_permission(self, authorizer_name, authorizer_lambda_function_arn):
+        """Constructs and returns the Lambda Permission resource allowing the Authorizer to invoke the function.
+
+        :returns: the permission resource
+        :rtype: model.lambda_.LambdaPermission
+        """
+        rest_api = ApiGatewayRestApi(self.logical_id, depends_on=self.depends_on)
+        api_id = rest_api.get_runtime_attr('rest_api_id')
+
+        partition = ArnGenerator.get_partition_name()
+        resource = '${__ApiId__}/authorizers/*'
+        source_arn = fnSub(ArnGenerator.generate_arn(partition=partition, service='execute-api', resource=resource),
+                           {"__ApiId__": api_id})
+
+        lambda_permission = LambdaPermission(self.logical_id + authorizer_name + 'AuthorizerPermission')
+        lambda_permission.Action = 'lambda:invokeFunction'
+        lambda_permission.FunctionName = authorizer_lambda_function_arn
+        lambda_permission.Principal = 'apigateway.amazonaws.com'
+        lambda_permission.SourceArn = source_arn
+
+        return lambda_permission
+
+    def _construct_authorizer_lambda_permission(self):
+        auth_properties = AuthProperties(**self.auth)
+        authorizers = self._get_authorizers(auth_properties.Authorizers)
+
+        if not authorizers:
+            return []
+
+        permissions = []
+
+        for authorizer_name, authorizer in authorizers.items():
+            # Construct permissions for Lambda Authorizers only
+            if not authorizer.function_arn:
```

Comment:
Is this the best way to check if this is a lambda authorizer? Are lambda authorizers required to have an ARN at this point, and is it guaranteed that the other authorizer types won't?

Reply:
Yes they must have a FunctionArn and yes this is the best way to differentiate between Lambda and non-Lambda. If new authorizer types are added in the future which have a FunctionArn it is likely they will also need a permission, and on the off-chance they don't then we can update this logic.